src/rust/src/x509/verify.rs

```diff
@@ -0,0 +1,218 @@
+    let webpki = x509_module
+        .getattr(pyo3::intern!(py, "Profile"))?
+        .get_item(pyo3::intern!(py, "WebPKI"))?;
+
+    let name = match name {
+        Some(name) => Some(
+            DNSString::new(name.to_str()?)
+                .ok_or_else(|| pyo3::exceptions::PyValueError::new_err("invalid DNS name"))?,
+        ),
+        None => None,
+    };
+
+    let time = time.map(|time| py_to_datetime(py, time)).transpose()?;
+
+    if profile.eq(rfc5280)? {
+        Ok(PyPolicy(FixedPolicy::RFC5280(Policy::new(name, time))))
+    } else if profile.eq(webpki)? {
+        Ok(PyPolicy(FixedPolicy::WebPKI(Policy::new(name, time))))
+    } else {
+        Err(pyo3::exceptions::PyValueError::new_err(
+            "invalid profile specified; expected a Profile variant",
+        ))
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+// #[pyo3(signature = (leaf, policy, intermediates, store))]
+fn verify<'p>(
+    py: pyo3::Python<'p>,
+    leaf: &PyCertificate,
+    policy: &PyPolicy,
+    intermediates: &'p pyo3::types::PyList,
+    store: &'p pyo3::PyAny,
+) -> CryptographyResult<Vec<PyCertificate>> {
+    let intermediates = intermediates
+        .iter()
+        .map(|o| o.extract::<pyo3::PyRef<'p, PyCertificate>>())
+        .collect::<Result<Vec<_>, _>>()?;
+    let store_certs = store
+        .getattr(pyo3::intern!(py, "_certs"))?
+        .downcast::<pyo3::types::PyList>()?
+        .iter()
+        .map(|o| o.extract::<pyo3::PyRef<'p, PyCertificate>>())
+        .collect::<Result<Vec<_>, _>>()?;
+    let store = Store::new(store_certs.iter().map(|t| t.raw.borrow_dependent().clone()));
+
+    // TODO: This manual monomorphization over each policy type is pretty ugly;
+    // maybe fix it.
+    let chain = match &policy.0 {
+        FixedPolicy::RFC5280(policy) => cryptography_x509_validation::verify(
+            leaf.raw.borrow_dependent(),
+            intermediates
+                .iter()
+                .map(|i| i.raw.borrow_dependent().clone()),
+            policy,
+            &store,
+        ),
+        FixedPolicy::WebPKI(policy) => cryptography_x509_validation::verify(
+            leaf.raw.borrow_dependent(),
+            intermediates
+                .iter()
+                .map(|i| i.raw.borrow_dependent().clone()),
+            policy,
+            &store,
+        ),
+    }
+    .map_err(|e| pyo3::exceptions::PyValueError::new_err("validation failed: {e}"))?;
+
+    // TODO: Optimize this? Turning a Certificate back into a PyCertificate
+    // involves a full round-trip back through DER, which isn't ideal.
+    chain
+        .iter()
+        .map(|c| {
+            let raw = pyo3::types::PyBytes::new(py, &asn1::write_single(c)?);
+            Ok(PyCertificate {
+                raw: OwnedCertificate::try_new(raw.into(), |raw| {
+                    asn1::parse_single(raw.as_bytes(py))
+                })?,
+                cached_extensions: None,
+            })
+        })
+        .collect()
```

Comment:
Flagging for review: there might be a less wasteful way to do this round-trip; right now it makes at least two pretty big copies per certificate.